include/realtime_tools/realtime_barrier.hpp

```diff
@@ -0,0 +1,468 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * Publishing ROS messages is difficult, as the publish function is
+ * not realtime safe.  This class provides the proper locking so that
+ * you can call publish in realtime and a separate (non-realtime)
+ * thread will ensure that the message gets published over ROS.
+ *
+ * Author: Colin F. MacKenzie
+ */
+
+#ifndef REALTIME_TOOLS__REALTIME_BARRIER_HPP_
+#define REALTIME_TOOLS__REALTIME_BARRIER_HPP_
+
+#include <mutex>
+#include <thread>
+
+
+namespace realtime_tools
+{
+
+/// @brief Attempt a lock on a resource but fail if already locked
+class try_lock : public std::unique_lock<std::mutex>
+{
+public:
+  explicit try_lock(std::mutex & m)
+  : std::unique_lock<std::mutex>(m, std::try_to_lock)
+  {}
+};
+
+/// @brief Lock a resource or wait for it to be free
+class wait_lock : public std::unique_lock<std::mutex>
+{
+public:
+  explicit wait_lock(std::mutex & m)
+  : std::unique_lock<std::mutex>(m, std::defer_lock)
+  {
+    // will wait some time
+    while (!try_lock()) {
+      std::this_thread::sleep_for(std::chrono::microseconds(500));
+    }
+  }
+};
+
+/// @brief default options for realtime acess to resources
+struct realtime
+{
+  using lock_strategy = try_lock;
+};
+
+/// @brief default options for non-realtime access to resources
+struct non_realtime
+{
+  using lock_strategy = wait_lock;
+};
+
+
+/// @brief Implements the underlying mechanisms for communicating between threads.
+/// MemoryBarrier implements a double buffer swapping mechanism for interchanging data between
+/// threads. It allocates two T objects on the heap. When MemoryBarrier::swap() is called the two
+/// pointers are swapped. Generally one pointer is accessed from the non-realtime thread and the
+/// other is used by the realtime thread. Only one-way communication, in either direction, is
+/// possible. For two-way communication two separate MemoryBarriers can be used as a pair.
+///
+/// You should not use this class directly, you will work with ReadBarrier, WriteBarrier and
+/// DirectAccess classes only.
+///
+/// General strategy:
+///   For non-RT writing to RT thread, non-RT thread writes but doesnt swap, RT swaps and reads.
+///   For RT writing back to non-RT, RT writes but doesnt swap, NRT swaps and reads at will
+template<class T>
+class MemoryBarrier
+{
+public:
+  /// @brief Provides direct access to data
+  /// Should only be used from within the realtime thread. Default template args specify
+  /// DirectAccess should only try to lock the resource and on failure the DirectAccess object
+  /// would be equal to nullptr and therefore evaluate to false.
+  /// This is also used internally by the non-RT ReadBarrier and WriteBarrier classes but
+  /// implements the thread-safe access.
+  template<typename access_mode = realtime,
+    typename lock_strategy = typename access_mode::lock_strategy>
+  class DirectAccess : public lock_strategy
```

Comment:
The idea here is by default the usage choice of access_mode being non-realtime or realtime will decide what kind of lock_strategy is used (or perhaps the right word is recommended). That doesn't mean the caller can't force a different lock_strategy but this would be at their own peril in a sense. For example, if they choose a wait_lock when requesting realtime. I don't disagree with your logic about composition over inheritance, but if this was switched to composition, then the other classes referencing this would need to refer to lock_strategy as a member field and I believe the std::enable_if_t SFINAE may fail. Again, if you can present an alternative...